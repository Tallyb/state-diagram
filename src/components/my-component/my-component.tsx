import { Component, Prop, h, State, Method, JSX, Watch, ComponentWillLoad } from '@stencil/core';

export interface ITransition {
  name: string;
  to: string;
  description?: string;
}
export interface IState {
  name: string;
  description?: string;
  transitions: Array<ITransition>;
  isFinal?: boolean;
}

export interface IDiagram {
  [key: string]: IState
}

@Component({
  tag: 'my-component',
  styleUrl: 'my-component.css',
  shadow: true
})
export class MyComponent implements ComponentWillLoad{
  /**
   * State Diagram structure
   */
  @Prop() diagram: IDiagram;

  /**
   * The name of the initial state
   */
  @Prop() initial: string;

  @State() currentState: IState;
  
  @Method() 
  async getCurrentState() {
    return this.currentState;
  }

  @Watch('initial')
  initialHandler(newValue: string) {
    this.setState(newValue);
  }

  componentWillLoad() {
    this.setState(this.initial)
  }

  setState(name: string) {
    if (typeof this.diagram === 'object') {
      let newState = this.diagram[name];
      if (!newState) return;
      this.currentState = newState;

    }
  }

  renderTransitions(transitions: Array<ITransition>): JSX.Element {
    return (
    <ul class="transitions">
      {transitions.map(t => <li onClick={ (event: UIEvent) => this.setState(t.to)}>{t.name}</li>)}
    </ul>
  )};

  renderState(state): JSX.Element {
    return (
      <div class="state">
        <p>{state.name}</p>
        {this.renderTransitions(state.transitions)} 
      </div>
    )
  }

  render(): JSX.Element {

    return (
      <div class="container">
        {(!!this.currentState)
          ? this.renderState(this.currentState)
          : <div> State not yet known </div> 
        }
      </div>
    );
  }
}
