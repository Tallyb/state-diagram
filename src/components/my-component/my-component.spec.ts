import { newSpecPage } from '@stencil/core/testing';

import {MyComponent, IDiagram } from './my-component';



describe('my-component', () => {
  const diagram: IDiagram = {
    a: {name: 'a', description: 'aaa', transitions: [
      {name: 'toB', to: 'b'},
      {name: 'toC', to: 'c'}
    ]},
    b: {name: 'b', description: 'bbb', transitions: [
      {name: 'toA', to: 'a'},
      {name: 'toC', to: 'c'}
    ]},
    c: {name: 'c', description: 'ccc', transitions: [
      {name: 'toA', to: 'a'},
      {name: 'toB', to: 'b'}
    ]}
  }

  it('should render initial state', async () => {
    const page = await newSpecPage({
      components: [MyComponent],
      html: `<my-component></my-component>`
    });
    page.rootInstance.diagram = diagram; 
    page.rootInstance.initial = 'a';
    await page.waitForChanges();
    expect(page.root).toMatchSnapshot();
  });

  it('should render empty state if initial does not exist', async () => {
    const page = await newSpecPage({
      components: [MyComponent],
      html: `<my-component></my-component>`
    });
    page.rootInstance.diagram = diagram; 
    page.rootInstance.initial = 'x';
    await page.waitForChanges();
    expect(page.root).toMatchSnapshot();
  });


  fit('should move to a new State', async () => {
    const page = await newSpecPage({
      components: [MyComponent],
      html: `<my-component></my-component>`
    });
    page.rootInstance.diagram = diagram; 
    page.rootInstance.initial = 'a';
    await page.waitForChanges();
    let transitions = await page.root.shadowRoot.querySelectorAll('li');
    transitions[0].click();
    await page.waitForChanges();
    
    expect(await page.root.shadowRoot.querySelector('p').textContent).toEqual('b');

  });


});
