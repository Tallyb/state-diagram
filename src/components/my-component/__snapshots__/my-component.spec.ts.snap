// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`my-component should render empty state if initial does not exist 1`] = `
<my-component>
  <mock:shadow-root>
    <div class="container">
      <div>
        State not yet known
      </div>
    </div>
  </mock:shadow-root>
</my-component>
`;

exports[`my-component should render initial state 1`] = `
<my-component>
  <mock:shadow-root>
    <div class="container">
      <div class="state">
        <p>
          a
        </p>
        <ul class="transitions">
          <li>
            toB
          </li>
          <li>
            toC
          </li>
        </ul>
      </div>
    </div>
  </mock:shadow-root>
</my-component>
`;
